Documentation/design-docs/eh-writethru.md

```diff
@@ -0,0 +1,378 @@
+# Exception Handling Write Through Optimization.
+
+Write through is an optimization done on local variables that live across exception handling flow like a handler, filter, or finally so that they can be enregistered - treated as a register candidate - throughout a method.  For each variable live across one of these constructs, the minimum requirement is that a store to the variables location on the stack is placed between a reaching definition and any point of control flow leading to the handler, as well as a load between any return from a filter or finally and an upward exposed use.  Conceptually this maintains the value of the variable on the stack across the exceptional flow which would kill any live registers.  This transformation splits a local variable into multiple enregisterable compiler temporaries backed by the local variable on the stack. For local vars that additionally have appearances within a eh construct, a load from the stack local is inserted to a temp that will be enregistered within the handler.
+
```

Comment:
@sdmaclea 

I'm not saying stores are free, just that in the current model we do many loads and some stores
and in the new model (assuming the RA behaves sensibly) we'll do many fewer loads and about the same number of stores, and given their relative costs, that's why we expect this transformation to capture most of the benefit without adding a lot of complexity -- eg complexity that would come by trying to optimize away some or all of the stores.

Your comment reminds me there's another implicit bit of information that Russ should surface here: in the current runtime, all the allocatable registers are killed by exception flow.

As for deferring the spilling stores until exceptions happen, it seems viable but it might mean we need even more jit-generated side data (though possibly foldable with the current GC info). If we are going to go down this route we should also weigh it versus alternatives that have similar cost. For instance we could revise the contract with the runtime so that some allocatables are preserved across exception flow.